web-common/src/features/charts/render/utils.ts

```diff
@@ -0,0 +1,170 @@
+import equal from "fast-deep-equal";
+import type { EmbedOptions, VisualizationSpec } from "vega-embed";
+import { vega } from "vega-embed";
+import type { SignalListeners, View } from "./types";
+
+export function updateMultipleDatasetsInView(
+  view: View,
+  data: Record<string, unknown>,
+): void {
+  for (const [name, value] of Object.entries(data)) {
+    const getType = {};
+    if (value) {
+      if (!!value && getType.toString.call(value) === "[object Function]") {
+        const parsedValue = value as (dataset: unknown) => unknown;
+        parsedValue(view.data(name));
+      } else {
+        view.change(
+          name,
+          vega
+            .changeset()
+            .remove(() => true)
+            .insert(value),
+        );
+      }
+    }
+  }
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function shallowEqual<T extends Record<string, any>>(
+  a: T = {} as T,
+  b: T = {} as T,
+  ignore: Set<string> = new Set(),
+): boolean {
+  const aKeys = Object.keys(a);
+  const bKeys = Object.keys(b);
+
+  return (
+    a === b ||
+    (aKeys.length === bKeys.length &&
+      aKeys.filter((k) => !ignore.has(k)).every((key) => a[key] === b[key]))
+  );
+}
+
+export function removeSignalListenersFromView(
+  view: View,
+  signalListeners: SignalListeners,
+): boolean {
+  const signalNames = Object.keys(signalListeners);
+  for (const signalName of signalNames) {
+    try {
+      view.removeSignalListener(signalName, signalListeners[signalName]);
+    } catch (error) {
+      // eslint-disable-next-line no-console
+      console.warn("Cannot remove invalid signal listener.", error);
+    }
+  }
+
+  return signalNames.length > 0;
+}
+
+export function addSignalListenersToView(
+  view: View,
+  signalListeners: SignalListeners,
+): boolean {
+  const signalNames = Object.keys(signalListeners);
+
+  for (const signalName of signalNames) {
+    try {
+      view.addSignalListener(signalName, signalListeners[signalName]);
+    } catch (error) {
+      console.warn("Cannot add invalid signal listener.", error);
+    }
+  }
+
+  return signalNames.length > 0;
+}
+
+export function getUniqueFieldNames(specs: VisualizationSpec[]): Set<string> {
+  return new Set(specs.flatMap((o) => Object.keys(o)));
+}
+
+interface SpecChanges {
+  width: false | number;
+  height: false | number;
+  isExpensive: boolean;
+}
+
+export function computeSpecChanges(
```

Comment:
Can you add some comments to this function? E.g. Why do we do this? What classifies a change as expensive?